src/_pytest/config/findpaths.py

```diff
@@ -22,44 +23,109 @@ def exists(path, ignore=OSError):
         return False
 
 
-def getcfg(args, config=None):
+def _parse_ini_config(path: py.path.local) -> py.iniconfig.IniConfig:
+    """Parses the given generic '.ini' file using legacy IniConfig parser, returning
+    the parsed object.
+
+    Raises UsageError if the file cannot be parsed.
+    """
+    try:
+        return py.iniconfig.IniConfig(path)
+    except py.iniconfig.ParseError as exc:
+        raise UsageError(str(exc))
+
+
+def _get_ini_config_from_pytest_ini(path: py.path.local) -> Optional[Dict[str, Any]]:
+    """Parses and validates a 'pytest.ini' file.
+
+    If present, 'pytest.ini' files are always considered the source of truth of pytest
+    configuration, even if empty or without a "[pytest]" section.
+    """
+    iniconfig = _parse_ini_config(path)
+    if "pytest" in iniconfig:
+        return dict(iniconfig["pytest"].items())
+    else:
+        return {}
+
+
+def _get_ini_config_from_tox_ini(path: py.path.local) -> Optional[Dict[str, Any]]:
+    """Parses and validates a 'tox.ini' file for pytest configuration.
+
+    'tox.ini' files are only considered for pytest configuration if they contain a "[pytest]"
+    section.
+    """
+    iniconfig = _parse_ini_config(path)
+    if "pytest" in iniconfig:
+        return dict(iniconfig["pytest"].items())
+    else:
+        return None
+
+
+def _get_ini_config_from_setup_cfg(path: py.path.local) -> Optional[Dict[str, Any]]:
+    """Parses and validates a 'setup.cfg' file for pytest configuration.
+
+    'setup.cfg' files are only considered for pytest configuration if they contain a "[tool:pytest]"
+    section.
+
+    If a setup.cfg contains a "[pytest]" section, we raise a failure to indicate users that
+    plain "[pytest]" sections in setup.cfg files is no longer supported (#3086).
+    """
+    iniconfig = _parse_ini_config(path)
+
+    if "tool:pytest" in iniconfig.sections:
+        return dict(iniconfig["tool:pytest"].items())
+    elif "pytest" in iniconfig.sections:
+        fail(CFG_PYTEST_SECTION.format(filename="setup.cfg"), pytrace=False)
+    return None
+
+
+def _get_ini_config_from_pyproject_toml(
+    path: py.path.local,
+) -> Optional[Dict[str, Any]]:
+    """Parses and validates a ``pyproject.toml`` file for pytest configuration.
+
+    The ``[tool.pytest]`` table is used by pytest. If the file contains that section,
+    it is used as the config file.
+    """
+    import toml
+
+    config = toml.load(path)
+
+    result = config.get("tool", {}).get("pytest", {}).get("ini_options", None)
+    if result is not None:
+        # convert all scalar values to strings for compatibility with other ini formats
+        # conversion to actual useful values is made by Config._getini
+        def make_scalar(v):
+            return v if isinstance(v, (list, tuple)) else str(v)
+
+        return {k: make_scalar(v) for k, v in result.items()}
+    else:
+        return None
+
+
+def getcfg(args):
```

Comment:
Would be great if you're able to type-annotate this one :grin: Usually I don't ask this but in this case I think it can really be beneficial for clarifying things.

Reply:
I gave it a go there, let me know if you have any suggestions however. 👍